### PR TITLE
Update nullability annotations

### PR DESCRIPTION
**Pull request by drewnoakes**

### CODE REVIEW

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/GraphNodes/ViewProviders/GraphViewProviderBase.cs [hunk 1]
```diff
@@ -121,7 +122,7 @@ IEnumerable<string> GetExistingChildren()
         public virtual bool MatchSearchResults(
             IDependency topLevelDependency,
             Dictionary<string, HashSet<IDependency>> searchResultsPerContext,
-            out HashSet<IDependency>? topLevelDependencyMatches)
+            [MaybeNullWhen(returnValue: false)] out HashSet<IDependency>? topLevelDependencyMatches)
```
Comment (davidwengier) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/GraphNodes/ViewProviders/GraphViewProviderBase.cs]: Does this mean that the out param definitely won't be null when the method returns true?

Comment (drewnoakes) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/GraphNodes/ViewProviders/GraphViewProviderBase.cs]: The above construct is incorrect. Pushed a correction.

The four attributes (AllowNull/DisallowNull/NotNull/MaybeNull) and the conditional form above all override the type's declared nullability. So in the above example the HashSet should be non-null with the attribute specifying a null return when false (or a nullable type with NotNullWhen(true)).

Comment (davidwengier) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/GraphNodes/ViewProviders/GraphViewProviderBase.cs]: Yep, the update makes much more sense. Couldn't work out what "maybe null" meant when the parameter type was already nullable.

**Comment by drewnoakes** [comment]
I think this needs a newer compiler. I've looked but can't see how to update the reference.

**Comment by davkean** [comment]
`<MicrosoftNetCompilersVersion>` in Versions.props controls this.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/GraphNodes/ViewProviders/IDependenciesGraphViewProvider.cs [hunk 2]
```diff
@@ -60,6 +60,6 @@ internal interface IDependenciesGraphViewProvider
         bool MatchSearchResults(
             IDependency topLevelDependency,
             Dictionary<string, HashSet<IDependency>> searchResultsPerContext,
-            [NotNullWhenTrue] out HashSet<IDependency>? topLevelDependencyMatches);
+            [MaybeNullWhen(false)] out HashSet<IDependency> topLevelDependencyMatches);
```
Comment (davkean) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/GraphNodes/ViewProviders/IDependenciesGraphViewProvider.cs]: If it "may be" null, then why isn't it marked nullable?

Comment (drewnoakes) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/GraphNodes/ViewProviders/IDependenciesGraphViewProvider.cs]: It actuall was until my force push. You use annotations when you want to override the declared type's nullability. Another example would be a `string?` property where the setter cannot be null (i.e. annotate the setter with `[DisallowNull]`). It's also useful on generic types such as `IEquatable<T>` (which reminds me, I should update one of the annotations for that interface in the dependencies node.)

**Comment by tmeschter** [comment]
We need to be careful that we don't conflict with #4929 as that also updates the Microsoft.Net.Compiler package.

**Pull request merged**